### PR TITLE
Adds support for subtracting NonNegatives and Positives from UBC annotations

### PR DESCRIPTION
**Pull request by kelloggm**

### CODE REVIEW

File: checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -460,10 +467,15 @@ public Void visitBinary(BinaryTree tree, AnnotatedTypeMirror type) {
...
                 case PLUS:
-                case MINUS:
                     // Dataflow refines this type if possible
                     type.addAnnotation(UNKNOWN);
                     break;
+                case MINUS:
+                    // Dataflow refines this type most of the time.
+                    // But when a NonNegative is subtracted from a LTL,
+                    // the result is still LTL without an offset.
+                    addAnnotationForMinus(left, right, type);
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java]: This needs to be done in dataflow or else information about offsets will be lost.  See test cases I added: https://github.com/typetools/checker-framework/pull/1068/commits/aaa2cabd04d9e47a0b867e369cb9cb652d28dfca

**Comment by kelloggm** [comment]
I've made the changes you requested, especially moving the code into Transfer. It fixes the new tests you provided, and also fixes at least one other test I discovered. I think it should be ready to merge modulo the tests passing, and the zero-expressions you merged in; not sure why you did that, though.

**Comment by smillst** [comment]
> the zero-expressions you merged in; not sure why you did that, though.

Because I wanted to use "0" as an offset in my example, but "0" wasn't recognized as an int literal.  That bug is fixed is PR #1071.  

I'll review these changes. Thanks!

**Comment by smillst** [comment]
This change be merged once all the tests pass and PR #1071 has been merged.

File: framework/src/org/checkerframework/framework/util/FlowExpressionParseUtil.java [hunk 2]
```diff
@@ -100,9 +100,9 @@ private static Pattern anchored(String regex) {
     /** Matches a field access. Capturing groups 1 and 2 are the object and field. */
     protected static final Pattern memberselect = anchored("([^.]+)\\.(.+)");
     /** Matches integer literals */
-    protected static final Pattern intPattern = anchored("[-+]?[1-9][0-9]*");
+    protected static final Pattern intPattern = anchored("0|[-+]?[1-9][0-9]*");
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/util/FlowExpressionParseUtil.java]: Please undo the changes to this line and `longPattern`.

File: framework/tests/flowexpression/ValueLiterals.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import testlib.flowexpression.qual.FlowExp;
```
Comment (mernst) [inline framework/tests/flowexpression/ValueLiterals.java]: This looks like it shouldn't be showing up as a diff in your pull request, either.

File: checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java [hunk 4]
```diff
@@ -303,14 +303,23 @@ private void refineNeqArrayLength(
     /**
      * If some Node a is known to be less than the length of some array, x, then, the type of a - b,
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java]: Delete commas after "array", "then", and "b".

File: checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java [hunk 5]
```diff
@@ -303,14 +303,23 @@ private void refineNeqArrayLength(
...
      * is @LTLengthOf(value="x", offset="b"). If b is known to be less than the length of some other
```
Comment (mernst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java]: Should the offset be `-b`?  I would think that "a+b" is `@LtLengthOf("x", "b")`.

**Comment by mernst** [comment]
There are conflicts that need to be resolved.
Also, something about this pull requests seems to be making all the Travis tests time out.

File: checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java (hunk 5 above)
Comment (smillst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java]: No, if `a < x.length`, then `a - b` is less than x.length if `b` is added to `a - b` and thus the offset is `+b`.

Comment (mernst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java]: What is the semantics of `@LtLengthOf` with an offset?  It's not documented, so I don't have a way of interpreting whether a given expression has that type.

Comment (smillst) [inline checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java]: I added documentation to LTLengthOf in PR #1073.
